pkg/kubelet/deviceplugin/endpoint.go

```diff
@@ -0,0 +1,195 @@
+
+	cancel context.CancelFunc
+	ctx    context.Context
+}
+
+func newEndpoint(socketPath, resourceName string, callback MonitorCallback) (*endpoint, error) {
+	client, err := dial(socketPath)
+	if err != nil {
+		glog.Errorf("Can't create new endpoint with path %s err %v", socketPath, err)
+		return nil, err
+	}
+
+	ctx, stop := context.WithCancel(context.Background())
+
+	return &endpoint{
+		client: client,
+
+		socketPath:   socketPath,
+		resourceName: resourceName,
+
+		devices:  nil,
+		callback: callback,
+
+		cancel: stop,
+		ctx:    ctx,
+	}, nil
+}
+
+func (e *endpoint) list() (pluginapi.DevicePlugin_ListAndWatchClient, error) {
+	glog.V(3).Infof("Starting List")
+	stream, err := e.client.ListAndWatch(e.ctx, &pluginapi.Empty{})
+	if err != nil {
+		glog.Errorf(ErrListAndWatch, e.resourceName, err)
+
+		return nil, err
+	}
+
+	devs, err := stream.Recv()
+	if err != nil {
+		glog.Errorf(ErrListAndWatch, e.resourceName, err)
+		return nil, err
+	}
+
+	devices := make(map[string]*pluginapi.Device)
+	var added, updated, deleted []*pluginapi.Device
+	for _, d := range devs.Devices {
+		devices[d.ID] = d
+		added = append(added, CloneDevice(d))
+	}
+
+	e.mutex.Lock()
+	e.devices = devices
+	e.mutex.Unlock()
+
+	e.callback(e.resourceName, added, updated, deleted)
+
+	return stream, nil
+}
+
+func (e *endpoint) listAndWatch(stream pluginapi.DevicePlugin_ListAndWatchClient) {
+	glog.V(3).Infof("Starting ListAndWatch")
+
+	devices := make(map[string]*pluginapi.Device)
+
+	e.mutex.Lock()
+	for _, d := range e.devices {
+		devices[d.ID] = CloneDevice(d)
+	}
+	e.mutex.Unlock()
+
+	for {
+		response, err := stream.Recv()
+		if err != nil {
+			glog.Errorf(ErrListAndWatch, e.resourceName, err)
+			return
+		}
+
+		devs := response.Devices
+		glog.V(2).Infof("State pushed for device plugin %s", e.resourceName)
+
+		newDevs := make(map[string]*pluginapi.Device)
+		var added, updated []*pluginapi.Device
+
+		for _, d := range devs {
+			dOld, ok := devices[d.ID]
+			newDevs[d.ID] = d
+
+			if !ok {
+				glog.V(2).Infof("New device for Endpoint %s: %v", e.resourceName, d)
+
+				devices[d.ID] = d
+				added = append(added, CloneDevice(d))
+
+				continue
+			}
+
+			if d.Health == dOld.Health {
+				continue
+			}
+
+			if d.Health == pluginapi.Unhealthy {
+				glog.Errorf("Device %s is now Unhealthy", d.ID)
+			} else if d.Health == pluginapi.Healthy {
+				glog.V(2).Infof("Device %s is now Healthy", d.ID)
+			}
+
+			devices[d.ID] = d
+			updated = append(updated, CloneDevice(d))
+		}
+
+		var deleted []*pluginapi.Device
+		for id, d := range devices {
+			if _, ok := newDevs[id]; ok {
+				continue
+			}
+
+			glog.Errorf("Device %s was deleted", d.ID)
+
+			deleted = append(deleted, CloneDevice(d))
+			delete(devices, id)
+		}
+
+		e.mutex.Lock()
+		e.devices = devices
+		e.mutex.Unlock()
+
+		e.callback(e.resourceName, added, updated, deleted)
+	}
+
+}
+
+func (e *endpoint) allocate(devs []string) (*pluginapi.AllocateResponse, error) {
```

Comment:
@RenaudWasTaken @jiayingz Pod details should also be passed along with device ids to the plugin. One use case for this requirement is special NIC resources such as Solarflare and Infiniband. With current API, it is possible to mount device files and libs inside container but the corresponding network interface will be available to use inside container only if `--net=host` is used. This is kinda very strict limitation.
Instead, if pod details are also passed to the device plugin, device plugin can pass the allocated interface from  the host net namespace to container net namespace and thus making it possible to use such pass through network interfaces inside non net=host containers as well.
NICs are one such use case, in future there could be many more. We should start with a wider scope and pass pod details as well to the device plugins and depending on requirement, plugins may or may not use these details. Thoughts?
cc @vishh @jeremyeder @sjenning @cmluciano

Reply:
@vikaschoudhary16 the unofficial, unvetted plan is to have device plugins drop a CNI config along with necessary CNI utility binaries, and have device plugins return additional interface via CNI constructs. That way the device plugin API can be invoked during pod creation rather than during container creation. 
In the future, if the networking SIG ends up having a similar RPC based CNI API, then the workflow will become simpler.

Reply:
The unofficial plan sounds good to me on the surface. I think @3XX0 and I determined that an  CNI plugin for Infiniband would probably be necessary.

Reply:
Agree. I think the discussion on CNI integration will probably wait after 1.8.